### PR TITLE
[Misc] Dragonwell environment variables

### PR DESCRIPTION
**Pull request by joeyleeeeeee97**
Summary: DRAGONWELL_JAVA_TOOL_OPTIONS(Dragonwell vendor specific java
options), DRAGONWELL_JAVA_TOOL_OPTIONS_JDK_ONLY(ignore JAVA_TOOL_OPTIONS
except for java command)

Test Plan: TestVendorVMOptions.java

Reviewed-by: yuleil, D-D-H

Issue: https://github.com/alibaba/dragonwell8/issues/330

### CODE REVIEW

**Review by D-D-H — CHANGES_REQUESTED** [review]
A quick comment:
Some extra changes should be made, such as vmError.cpp(it will print the environment vars).

**Comment by yuleil** [comment]
Can it be implemented like this？

- VM only recognizes `DRAGONWELL_JAVA_OPTIONS`
- and if `DRAGONWELL_JAVA_TOOL_OPTIONS_JDK_ONLY` is recognized in `java.c`, it will be automatically appended to `DRAGONWELL_JAVA_OPTIONS`

**Comment by joeyleeeeeee97** [comment]
@D-D-H 
Added in vmErrors.cpp.
Skipped for jvmti.xml,  supporting these Dragonwell options for  JNI_CreateJavaVM seems unnecessary..

image

File: hotspot/src/share/vm/utilities/vmError.cpp [hunk 1]
```diff
@@ -54,7 +54,7 @@ PRAGMA_FORMAT_MUTE_WARNINGS_FOR_GCC
 const char *env_list[] = {
   // All platforms
   "JAVA_HOME", "JRE_HOME", "JAVA_TOOL_OPTIONS", "_JAVA_OPTIONS", "CLASSPATH",
-  "JAVA_COMPILER", "PATH", "USERNAME",
+  "JAVA_COMPILER", "PATH", "USERNAME", "DRAGONWELL_JAVA_TOOL_OPTIONS", "DRAGONWELL_JAVA_TOOL_OPTIONS_JDK_ONLY"
```
Comment (D-D-H) [inline hotspot/src/share/vm/utilities/vmError.cpp:57]: missing ',' at the end

File: hotspot/src/share/vm/runtime/arguments.cpp [hunk 2]
```diff
@@ -3766,9 +3779,22 @@ jint Arguments::parse_java_options_environment_variable(SysClassPath* scp_p, boo
...
+      return parse_options_environment_variable("JAVA_TOOL_OPTIONS", scp_p,
+                                                  scp_assembly_required_p);
+    }
+    return JNI_OK;
+  }
   return parse_options_environment_variable("JAVA_TOOL_OPTIONS", scp_p,
                                             scp_assembly_required_p);
 }
+jint Arguments::parse_dragonwell_options_environment_variable(SysClassPath* scp_p, bool* scp_assembly_required_p) {
+  return parse_options_environment_variable("DRAGONWELL_JAVA_TOOL_OPTIONS", scp_p,
```
Comment (D-D-H) [inline hotspot/src/share/vm/runtime/arguments.cpp:3795]: why not use parse_options_environment_variable directly?

File: hotspot/src/share/vm/utilities/vmError.cpp (hunk 1 above)
Comment (joeyleeeeeee97) [reply hotspot/src/share/vm/utilities/vmError.cpp:57]: done

File: hotspot/src/share/vm/runtime/arguments.cpp (hunk 2 above)
Comment (joeyleeeeeee97) [reply hotspot/src/share/vm/runtime/arguments.cpp:3795]: image
just like JAVA_TOOL_OPTIONS..

File: hotspot/src/share/vm/utilities/vmError.cpp (hunk 1 above)
Comment (D-D-H) [reply hotspot/src/share/vm/utilities/vmError.cpp:57]: is it really done?

File: hotspot/src/share/vm/runtime/arguments.cpp [hunk 3]
```diff
@@ -932,6 +933,9 @@ bool Arguments::process_argument(const char* arg,
       jio_fprintf(defaultStream::error_stream(), "%s", locked_message_buf);
     }
   } else {
+    if (ignore_unrecognized) {
```
[the remark this answers is not in the document]
Comment (D-D-H) [reply hotspot/src/share/vm/runtime/arguments.cpp:936]: hmm... If it is an upstream problem, why include this fix in this patch?

[the remark this answers is not in the document]
Comment (joeyleeeeeee97) [reply hotspot/src/share/vm/runtime/arguments.cpp:936]: customer requires this function in production, might takes too long to submit to upstream and merge back

File: jdk/src/share/bin/java.c [hunk 4]
```diff
@@ -1237,6 +1239,7 @@ InitializeJVM(JavaVM **pvm, JNIEnv **penv, InvocationFunctions *ifn)
                    i, args.options[i].optionString);
     }
 
+    ifn->SetRunningFromJava(!printVersion&&runFromJava);
```
[the remark this answers is not in the document]
Comment (joeyleeeeeee97) [reply jdk/src/share/bin/java.c:1242]: arguments are parsed at JNI level, we need to identify whether it is running from `java` and set the state to vm via jni.

File: jdk/src/share/bin/java.c [hunk 5]
```diff
@@ -203,6 +204,7 @@ JLI_Launch(int argc, char ** argv,              /* main argc, argc */
     _is_java_args = javaargs;
     _wc_enabled = cpwildcard;
     _ergo_policy = ergo;
+    runFromJava = jargv == NULL;
```
[the remark this answers is not in the document]
Comment (joeyleeeeeee97) [reply jdk/src/share/bin/java.c:207]: jargv and jargc is parsed from const_extra_jargs, which controlled by EXTRA_JAVA_ARGS during compilation, for java command EXTRA_JAVA_ARGS is never defined so this is NULL. (for tools this must be defined)

As in dragonwell 11.
    JLI_InitArgProcessing(jargc > 0, const_disable_argfile);

https://github.com/alibaba/dragonwell11/blob/1529c150d3a34b6b01d384147c9d3fe3be1c73ea/src/java.base/share/native/launcher/main.c#L141

File: hotspot/src/share/vm/utilities/vmError.cpp (hunk 1 above)
Comment (joeyleeeeeee97) [reply hotspot/src/share/vm/utilities/vmError.cpp]: yes!..

**Review by yuleil — APPROVED** [review]

File: hotspot/src/share/vm/runtime/arguments.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (D-D-H) [reply hotspot/src/share/vm/runtime/arguments.cpp:936]: Please add more related comments to the git commit if this change must be included in this patch.

File: hotspot/src/share/vm/runtime/arguments.cpp (hunk 2 above)
Comment (D-D-H) [reply hotspot/src/share/vm/runtime/arguments.cpp:3795]: okay

File: hotspot/src/share/vm/runtime/arguments.cpp [hunk 6]
```diff
@@ -3458,6 +3468,9 @@ jint Arguments::parse_each_vm_init_arg(const JavaVMInitArgs* args,
     } else if (match_jfr_option(&option)) {
       return JNI_EINVAL;
 #endif
+    } else if (match_option(option, "-XX:+IgnoreUnrecognizedVMOptions", &tail)) {
```
[the remark this answers is not in the document]
Comment (D-D-H) [reply hotspot/src/share/vm/runtime/arguments.cpp:3471]: Please add more related comments to the git commit if this change must be included in this patch.

**Comment by D-D-H** [comment]
I still don't think it's a good idea to include IgnoreUnrecognizedVMOptions fix change in this patch.
@yuleil What do you think?

**Review by D-D-H — APPROVED** [review]

**Pull request merged**